### PR TITLE
Support  ONNX Op with variadic output using Krnl.Call

### PR DESCRIPTION
**Pull request by chentong319**
krnl.call take the outputs and inputs of an ONNX Op as its parameters. A new attribute records the number of outputs. This attribute is needed to set the side effect correctly. The builder to build krnl.call from an ONNX Op are updated to the change.

### CODE REVIEW

File: src/Dialect/Krnl/Krnl.td [hunk 1]
```diff
@@ -52,35 +52,51 @@ def KrnlCallOp : Op<Krnl_Dialect, "call",
...
-    OMTensor is used as a container for MemRef arguments.
+    OMTensor is used as a container for MemRef arguments. Other representation
+    of parameters, such as data pointer only, will be supported in future.
   }];
 
-  let arguments = (ins StrAttr:$funcName, AnyType:$result, Variadic<AnyType>:$parameters);
+  let arguments = (ins StrAttr:$funcName,
+      // Should mlir::OpTrait::AttrSizedOperandSegments be used?
+      // The case for krnl.Call is simple.
+      DefaultValuedAttr<SI64Attr, "1">:$num_of_output,
```
Comment (tungld) [inline src/Dialect/Krnl/Krnl.td:86]: Naming issue: please change `num_of_output` to `numOfOutput` to make it consistent with `funcName`.

**Review by tungld — COMMENTED** [review]
Could you add a lit test about lowering multi-output Krnl.Call to LLVM? Thanks!

**Comment by chentong319** [comment]
The lit test for ONNX to Krnl may be added by Yashushi because he has an use case of the new functionality. The other change of this PR is Krnl to LLVM. We do not have lit test for that pass.

**Comment by chentong319** [comment]
There are already some lit tests for krill.call. The new builder for multiple outputs work for them.

**Comment by negiyas** [comment]
> The lit test for ONNX to Krnl may be added by Yashushi because he has an use case of the new functionality. The other change of this PR is Krnl to LLVM. We do not have lit test for that pass.

@chentong319 I am adding my cases as lit tests. Thank you.

**Comment by negiyas** [comment]
@chentong319 May I have one question about this PR?

I created the following lit tests for krnl-to-llvm conversion of krnl.Call op to compare the converted llvm results among the "numOutput=1," "numOutput=2," and "numOutput=3" cases, but I cannot see any difference among the converted results except the test case names.  Do the following lit tests make sense? 

```
// RUN: onnx-mlir-opt -O3 --convert-krnl-to-affine --convert-krnl-to-llvm="use-opaque-pointers=true" %s -split-input-file | FileCheck %s

// COM:
// COM: Convert krnl.call with one return value.
// COM: Expect that buffer pointed by the first argument is updated by the
// COM: testFunc function, and buffer pointed by %0 is updated and returned.
func.func @test_krnlcall_with_one_return_value() -> (memref<10x10xf32>, memref<20x20xf32>, memref<30x30xf32>) {
  %0 = memref.alloc() {alignment = 4096 : i64} : memref<10x10xf32>
  %1 = memref.alloc() {alignment = 4096 : i64} : memref<20x20xf32>
  %2 = memref.alloc() {alignment = 4096 : i64} : memref<30x30xf32>
  "krnl.call"(%0, %1, %2) {funcName= "testFunc", numOfOutput = 1 : si64} : (memref<10x10xf32>, memref<20x20xf32>, memref<30x30xf32>) -> ()
  return %0, %1, %2 : memref<10x10xf32>, memref<20x20xf32>, memref<30x30xf32>
}

// -----

// COM:
// COM: Convert krnl.call with two return values.
// COM: Expect that buffers pointed by the first and second arguments are
// COM: updated by the testFunc function, and buffers pointed by %0 and %1 are
// COM: updated and returned.
func.func @test_krnlcall_with_two_return_values() -> (memref<10x10xf32>, memref<20x20xf32>, memref<30x30xf32>) {
  %0 = memref.alloc() {alignment = 4096 : i64} : memref<10x10xf32>
  %1 = memref.alloc() {alignment = 4096 : i64} : memref<20x20xf32>
  %2 = memref.alloc() {alignment = 4096 : i64} : memref<30x30xf32>
  "krnl.call"(%0, %1, %2) {funcName= "testFunc", numOfOutput = 2 : si64} : (memref<10x10xf32>, memref<20x20xf32>, memref<30x30xf32>) -> ()
  return %0, %1, %2 : memref<10x10xf32>, memref<20x20xf32>, memref<30x30xf32>
}

// -----

// COM:
// COM: Convert krnl.call with three return values.
// COM: Expect that buffers pointed by the first, second and third arguments are
// COM: updated by the testFunc function, and buffers pointed by %0, %1 and %2
// COM: are updated and returned.
func.func @test_krnlcall_with_three_return_values() -> (memref<10x10xf32>, memref<20x20xf32>, memref<30x30xf32>) {
  %0 = memref.alloc() {alignment = 4096 : i64} : memref<10x10xf32>
  %1 = memref.alloc() {alignment = 4096 : i64} : memref<20x20xf32>
  %2 = memref.alloc() {alignment = 4096 : i64} : memref<30x30xf32>
  "krnl.call"(%0, %1, %2) {funcName= "testFunc", numOfOutput = 3 : si64} : (memref<10x10xf32>, memref<20x20xf32>, memref<30x30xf32>) -> ()
  return %0, %1, %2 : memref<10x10xf32>, memref<20x20xf32>, memref<30x30xf32>
}
```

**Review by tungld — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**